Add LoC(lines of code) tracker

willdavidson05 opened the pull request:
<!-- _modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_
_referenced with modifications from [pycytominer](https://github.com/cytomining/pycytominer/blob/master/.github/PULL_REQUEST_TEMPLATE.md)_ -->

# Description

<!--
Thank you for your contribution to the Software Gardening Almanack!

Please _succinctly_ summarize your proposed change.
Namely, consider addressing the following questions:

- What motivated you to open this pull request?
- Were there any special adjustments you had to make to complete the work?
- Are there any issues which are related to this pull request (you may used a `#<digit>` to reference GitHub issues as links within this description)?

Also, if you haven't already, please use `pre-commit run --all-files` to help check your files using this project's pre-commit configuration.
-->
This PR create a `LoC_tracker.py` module that retrieves file statistics from a specified source and target commit. From these statistics, we find the total number of insertions and deletions, sum them together, and then return the total.

A test was made `test_LoC.py`, that checks that the module works correctly.

I appreciate any comments or feedback!

Closes #54 

## What is the nature of your change?

- [x] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own contributions.
- [x] I have commented my content, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to related documentation (outside of book content).
- [x] My changes generate no new warnings.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests that prove my additions are effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

gwaybio commented on src/almanack/LoC_tracker.py:
```diff
@@ -0,0 +1,21 @@
+    for commit in repo.iter_commits():
+        # Retrieve commit statistics
+        diff_stat = commit.stats.total
+        lines_added = diff_stat["insertions"]
```
Are you envisioning that you'll do something differently with insertions and deletions in the future? If not, then why not simply append commit.stats.total?

willdavidson05 replied:
This is a good point! I did this initially for print statement testing, however I see how this is unnecessary . This helped me find the "lines" attribute of `stats` which works even better. Thank you for this!

gwaybio commented on src/almanack/LoC_tracker.py:
```diff
@@ -0,0 +1,21 @@
+import git
+
+
+def calculate_loc_changes(repo_path: str):
```
I recommend adding a docstring to this function. It's a good habit to get into!

gwaybio commented on tests/test_Loc.py:
```diff
@@ -0,0 +1,15 @@
+def test_calculate_loc_changes(repository_paths: dict[str, pathlib.Path]):
+    high_entropy_path = repository_paths["high_entropy"]
+    low_entropy_path = repository_paths["low_entropy"]
+
```
I'm not totally confident, but don't these require assertions to make a valid test?

willdavidson05 replied:
Whoops, you're totally right!  I have corrected this mistake and created a robust test case.

gwaybio reviewed (commented):
A couple comments. I'm general, great to see such comprehensive docs on the PR description and that it was a bite-sized change. Keep it up!

willdavidson05 commented:
Thank you for the review @gwaybio !! Your comments really helped me simplify/optimize my code and test cases. I have requested another review but no rush to get to it.

d33bs commented on src/almanack/LoC_tracker.py:
Consider generalizing the name for this module to help invite further work where / when needed.

willdavidson05 replied:
Will do! I have been thinking about a new name, and the best I have is `code_tracker.py`. Let me know what you think of this! I'm not sure how I feel about it yet.

d33bs commented on src/almanack/LoC_tracker.py:
```diff
@@ -0,0 +1,26 @@
+
+    total_lines_changed = 0
+
+    for commit in repo.iter_commits():
```
Would it be possible to tie in `almanack.git_parser.get_commit_logs` here somehow to help introduce flexibility when it comes to certain commit segments or time periods (instead of only in full summary)? Perhaps it could come in the form of an expansion of the data structure found within that function (another dictionary key-value pair, for example). For example, we might expect that earlier in a project's lifespan the amount of change will be higher than later on. These counts from the earlier time periods might inadvertently effect calculations for later time periods under certain circumstances.

willdavidson05 replied:
Thanks for the idea Dave! I was able to implement the get_commit_logs by adding in attributes the my function, but wanted to see what you think?

d33bs commented on tests/data/almanack/entropy/add_data.py:
```diff
@@ -44,8 +44,8 @@ def create_repositories(base_path: str):
 
-    # Run the add_entropy.py scriptwhat
-    add_entropy(base_path)
+    # Run the add_entropy.py script
```
Is `add_entropy.py` a script or a module? Consider revising this based on what you think.

d33bs commented on tests/data/almanack/entropy/add_entropy.py:
```diff
@@ -75,7 +75,7 @@ def add_entropy_to_file(file_path: str, lines_of_code: str):
 
 
-def add_entropy(base_path):
+def insert_entropy(base_path):
```
Consider adding a type hint for the parameter and return type.

d33bs commented on src/almanack/LoC_tracker.py:
```diff
@@ -0,0 +1,26 @@
+import git
+
+
+def calculate_loc_changes(repo_path: str):
```
Consider adding a type hint for the return type.

d33bs commented on src/almanack/LoC_tracker.py:
```diff
@@ -0,0 +1,26 @@
+    """
+    repo = git.Repo(repo_path)
+
+    total_lines_changed = 0
```
Consider the use of a [list comprehension](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions) to combine this line with the for loop to make the necessary calculations (this might depend on how you approach things given the below comment as well).

d33bs reviewed (commented):
Nice job! I left a few comments and suggestions. Please don't hesitate to reach out with any questions.

willdavidson05 commented:
Thank you @d33bs for the review! I have made some minor changes based off of your comments. I added attributes to my  `git_commit_logs` function so I am able to reference my `git_parser.py` module through `code_tracker.py`. I also added hint return types for every function I have done so far. I left a couple questions on your comments as well!

willdavidson05 commented:
@d33bs Wanted to give a couple notes about what I have changed. 
1.) I deleted the `get_all_commit_logs` function inside of `git_parser.py`. I no longer see the need for it at this time
2.) I changed the `calculate_loc_changes` function to accept two new arguments source and target. This allows a file-based approach to our calculation. 
3.) I created a function that grabs the two most recent commits(source,target) for the test datasets. I'm not sure if this is the best way to go about it, but would love to hear what you think! I also have it in `test_code_tracker.py` because I didn't see the utility of having it anywhere else.

d33bs commented on src/almanack/code_tracker.py:
```diff
@@ -0,0 +1,31 @@
+from .git_parser import get_commit_logs
+
+
+def calculate_loc_changes(repo_path: pathlib.Path, source: str, target: str) -> int:
```
Would it make sense to include this function within the `git_parser` module? I ask as it seems to use similar parameters / attributes. If you opt to include it in the same module, it could help with the organization of the package and development.

d33bs commented on src/almanack/code_tracker.py:
```diff
@@ -0,0 +1,31 @@
+        repo_path (pathlib.Path): The path to the git repository.
+        source (str): The source commit hash.
+        target (str): The target commit hash.
+    """
```
Consider adding a return type section to this docstring.

d33bs commented on src/almanack/code_tracker.py:
```diff
@@ -0,0 +1,31 @@
+
+def calculate_loc_changes(repo_path: pathlib.Path, source: str, target: str) -> int:
+    """
+    Finds the total number of code lines changed between the source and target commits.
```
Would this be able to parse individual file-level lines of code differences between the source and target commits? If not, consider expanding the function to include this capability (perhaps with an additional parameter which identifies a file within a tree) to address this.

d33bs commented on src/almanack/code_tracker.py:
```diff
@@ -0,0 +1,31 @@
+    target_commit_info = commit_logs[target]
+
+    # Calculate total lines changed between the source and target commits
+    total_lines_changed = (
+        source_commit_info["stats"]["total"]["lines"]
+        - target_commit_info["stats"]["total"]["lines"]
+    )
```
Consider returning the calculation directly without the use of an additional single-use named variable.

d33bs commented on tests/test_code_tracker.py:
```diff
@@ -0,0 +1,40 @@
+from almanack.git_parser import get_commit_logs
+
+
+def get_most_recent_commits(repo_path: pathlib.Path) -> tuple[str, str]:
```
Nice!

d33bs reviewed (commented):
Great work! I left a few comments for your consideration - interested in what you think about these. Please don't hesitate to let me know if you have any questions.

willdavidson05 commented:
@d33bs, thanks for the review! I have implemented most of your feedback. Just a comment about a question from our meeting earlier, I found that the filename is the filepath relative to the repository root. Thank you!

d33bs commented on src/almanack/git_parser.py:
```diff
@@ -49,19 +56,22 @@ def get_commit_contents(repository_path: str, commit_id: str):
+    diff = repo.git.diff(source, target, "--numstat")
+    return {
+        filename: abs(int(removed) + int(added))  # Calculate change
+        for added, removed, filename in (line.split() for line in diff.splitlines())
```
Given how you gather data from this (data within a string) would it make sense to expand the test data to ensure nothing unusual happens when you have more than one file in a repository? Mostly I wonder, what would happen if there were two files (perhaps with one in a subdir and the other not)?

d33bs reviewed (approved):
Nice job, this is looking pretty good to me! I left a couple comments about minor considerations; feel free to merge when you feel good about things.
